### PR TITLE
Add `aws_c_common` and `aws_lc`

### PR DESCRIPTION
**Pull request by giordano**

### CODE REVIEW

File: A/aws_c_common/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	..
+cmake --build . -j${nproc} --target install
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms()
+
+# The products that we will ensure are always built
+products = Product[
```
Comment (imciner2) [inline A/aws_c_common/build_tarballs.jl:42]: No products for these two?

Comment (giordano) [reply A/aws_c_common/build_tarballs.jl:42]: I don't know yet what they are :slightly_smiling_face:

**Comment by giordano** [comment]
@quinnj do I understand correctly aws_lc vendors openssl? :confused:

**Comment by quinnj** [comment]
> @quinnj do I understand correctly aws_lc vendors openssl? 😕

Hmmm, in their description, it says:

> AWS-LC is a general-purpose cryptographic library maintained by the AWS Cryptography team for AWS and their customers. It іs based on code from the Google BoringSSL project and the OpenSSL project.

So I thought it was a completely new project that copied some stuff, but not totally vendor?

**Comment by quinnj** [comment]
But I do see the `include/openssl` in the top-level directory, so maybe?

**Comment by giordano** [comment]
Looking at the installed files https://buildkite.com/julialang/yggdrasil/builds/5072#018a868f-d685-447a-9d89-e464000cf0f0/656-1201 sounds like they install `include/openssl`, and then the libraries are called `libcrypto` and `libssl`, which I presume are going to conflict with something else (unless we carefully do static linking)

**Comment by quinnj** [comment]
Yeah, I see. I don't think we need to expose lc or sn2 as shared libraries, only static for aws-c-io to use if that makes a difference. Ideally, we'd have aws-c-common, aws-c-io, aws-c-http, and maybe aws-c-cal as shared libraries that could be used stand-alone. s2n and lc are really just providing the implementations for the tlc_channel_handler in aws-c-io and the hashing in aws-c-cal, but I wouldnt' anticipate needing to us them on their own.

**Comment by giordano** [comment]
@quinnj what are your platforms requirements exactly?  AWSCRT wasn't built for Windows and FreeBSD (both of which are currently failing, plus some others :smiling_face_with_tear:)

**Comment by quinnj** [comment]
I was really just going for the big 3, no need for FreeBSD.

**Comment by giordano** [comment]
Do you confirm aws-lc isn't needed on macos and Windows: https://github.com/JuliaPackaging/Yggdrasil/blob/1521cc4b6c44f7dd52114436fcedc9e3b28176ec/A/AWSCRT/build_tarballs.jl#L29?

**Comment by quinnj** [comment]
Yeah, I believe I read that the libraries should pick up the system native TLS for windows and macOS, so I think that's right.

**Comment by quinnj** [comment]
Here's the snippet I saw that talked about the use of system native TLS
image
(from [here](https://github.com/awslabs/aws-c-io#special-pre-defined-handlers))

**Comment by giordano** [comment]
Ok, aws-lc now builds everywhere on Linux and FreeBSD.  We still have problems with aws-c-common on Windows: [32-bit Windows fails to build](https://buildkite.com/julialang/yggdrasil/builds/5098#018a8a9d-476c-4e2c-bd45-6e2540f653b7/635-1763):
```
[18:22:09] In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windows.h:92:0,
[18:22:09]                  from /workspace/srcdir/aws-c-common/source/allocator.c:15:
[18:22:09] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock.h: At top level:
[18:22:09] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock.h:286:37: error: conflicting types for ‘htonl’
[18:22:09]    WINSOCK_API_LINKAGE u_long WSAAPI htonl(u_long hostlong);
[18:22:09]                                      ^
[18:22:09] In file included from /workspace/srcdir/aws-c-common/include/aws/common/byte_order.h:70:0,
[18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/byte_buf.h:9,
[18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/thread.h:8,
[18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/logging.h:11,
[18:22:09]                  from /workspace/srcdir/aws-c-common/source/allocator.c:8:
[18:22:09] /workspace/srcdir/aws-c-common/include/aws/common/byte_order.inl:47:23: note: previous implicit declaration of ‘htonl’ was here
[18:22:09]      return ((uint64_t)htonl(low)) << 32 | htonl(high);
[18:22:09]                        ^
```
and actually also 64-bit is problematic, because it uses the functions `bcryptOpenAlgorithm` and `bcryptGenRandom`, which sounds somewhat similar to https://github.com/openssl/openssl/pull/13882.  @quinnj I don't know how much effort to put into this, is aws-c-common needed on Windows?

**Comment by quinnj** [comment]
> Ok, aws-lc now builds everywhere on Linux and FreeBSD. We still have problems with aws-c-common on Windows: [32-bit Windows fails to build](https://buildkite.com/julialang/yggdrasil/builds/5098#018a8a9d-476c-4e2c-bd45-6e2540f653b7/635-1763):
> 
> ```
> [18:22:09] In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windows.h:92:0,
> [18:22:09]                  from /workspace/srcdir/aws-c-common/source/allocator.c:15:
> [18:22:09] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock.h: At top level:
> [18:22:09] /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock.h:286:37: error: conflicting types for ‘htonl’
> [18:22:09]    WINSOCK_API_LINKAGE u_long WSAAPI htonl(u_long hostlong);
> [18:22:09]                                      ^
> [18:22:09] In file included from /workspace/srcdir/aws-c-common/include/aws/common/byte_order.h:70:0,
> [18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/byte_buf.h:9,
> [18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/thread.h:8,
> [18:22:09]                  from /workspace/srcdir/aws-c-common/include/aws/common/logging.h:11,
> [18:22:09]                  from /workspace/srcdir/aws-c-common/source/allocator.c:8:
> [18:22:09] /workspace/srcdir/aws-c-common/include/aws/common/byte_order.inl:47:23: note: previous implicit declaration of ‘htonl’ was here
> [18:22:09]      return ((uint64_t)htonl(low)) << 32 | htonl(high);
> [18:22:09]                        ^
> ```
> 
> ~and actually also 64-bit is problematic, because it uses the functions `bcryptOpenAlgorithm` and `bcryptGenRandom`, which sounds somewhat similar to [openssl/openssl#13882](https://github.com/openssl/openssl/pull/13882)~ _**Edit**_: I had messed up with the `sed` script.

I'm fine dropping support for windows 32-bit.

**Comment by giordano** [comment]
That error is fixed by https://github.com/awslabs/aws-c-common/pull/1059, but now I have something even more exciting:
```
In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/windows.h:71:0,
                 from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/winsock.h:10,
                 from /workspace/srcdir/aws-c-common/include/aws/common/byte_order.inl:14,
                 from /workspace/srcdir/aws-c-common/include/aws/common/byte_order.h:70,
                 from /workspace/srcdir/aws-c-common/include/aws/common/byte_buf.h:9,
                 from /workspace/srcdir/aws-c-common/include/aws/common/uri.h:8,
                 from /workspace/srcdir/aws-c-common/source/uri.c:5:
/workspace/srcdir/aws-c-common/source/uri.c:25:5: error: expected identifier before numeric constant
     ERROR,
     ^
```
Sounds like something is defining a macro called `ERROR`, screwing everything up :neutral_face:

**Comment by giordano** [comment]
I confirm my hypothesis:
```console
sandbox:${WORKSPACE}/srcdir/aws-c-common/build # /opt/bin/i686-w64-mingw32-libgfortran3-cxx11/i686-w64-mingw32-gcc --sysroot=/opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/ -DAWS_AFFINITY_METHOD=AWS_AFFINITY_METHOD_NONE -DCJSON_HIDE_SYMBOLS -DINTEL_NO_ITTNOTIFY_API -DWINDOWS_KERNEL_LIB=kernel32 @CMakeFiles/aws-c-common.dir/includes_C.rsp -D_WIN32_WINNT=0x0601 -O3 -DNDEBUG -Wall -Wstrict-prototypes -pedantic -Wno-long-long -D_FILE_OFFSET_BITS=64 -std=gnu99 -MD -MT CMakeFiles/aws-c-common.dir/source/uri.c.obj -MF CMakeFiles/aws-c-common.dir/source/uri.c.obj.d -E /workspace/srcdir/aws-c-common/source/uri.c -dM | grep '\<ERROR\>'
#define RGN_ERROR ERROR
#define ERROR 0
```
but I have no idea of where that `ERROR = 0` is coming from.

**Comment by quinnj** [comment]
https://github.com/tpn/winsdk-7/blob/f8a2d5f306b400f355ed3713b93725e6b8f69f1c/v7.1A/Include/WinGDI.h#L112?

**Comment by giordano** [comment]
Yes, I just got there :rage:

**Comment by giordano** [comment]
Finally all green :slightly_smiling_face:

**Comment by quinnj** [comment]
❤️ ❤️ ❤️

**Pull request merged**